hashes/src/impls.rs

```diff
@@ -44,6 +47,15 @@ impl_write!(
     |_us| { Ok(()) }
 );
 
+impl_write!(
+    sha384::HashEngine,
+    |us: &mut sha384::HashEngine, buf| {
+        us.input(buf);
+        Ok(buf.len())
+    },
+    |_us| { Ok(()) }
+);
```

Comment:
Why are these not called in the `hash_type` macro?

Reply:
Because almost nothing on `HashEngine` is macroized (neither are the `Default` impls or the `HashEngine` impl) (except for one function which is done by `engine_input_impl!`). We could, and probably should, change that, but I feel like that'd be a larger discussion of reorganizing the internal_macros in this crate which nobody really wants to do.

Also historically this file used to have explicit `io::Write` impls and the whole module was gated on `std`, which was "simple enough" not to bother macro-izing.

Reply:
> neither are the `Default` impls or the `HashEngine` impl

Those are hard to meaningfully deduplicate since the engines are different, although consistent calling of `new` from `Default` makes sense.

> We could, and probably should, change that

Yes, and I already did add stuff to `HashEngine` in #3130 

> but I feel like that'd be a larger discussion of reorganizing the internal_macro

I don't think there's anything to reorganize. The macro is supposed to implement standard stuff for our hash types, one call for each hash module and this is what it should continue to be. We might split up the default part to deduplicate siphash but that's a pretty simple refactor.

Reply:
I've a love hate relationship with macros at the moment, it feels awesome to write them and often painful AF to read and maintain them. I'll leave this PR as is for now but I'll keep it in mind.

Reply:
Well, the alternative is to maintain *other code* multiple times and make subtle mistakes doing it. So if you have to spend time on some BS and be annoyed by it then you might as well choose the BS that improves reliability.

Reply:
> So if you have to spend time on some BS and be annoyed by it then you might as well choose the BS that improves reliability.

Lolz, but a good perspective I suppose. I'll add macros to this PR.